### PR TITLE
Support restore from last savepoint recorded in job status

### PR DESCRIPTION
**Pull request by functicons**
This feature allows users to stop and resume jobs without explicitly
updating the savepoint to restore the job from. The controller
automatically uses the last savepoint recorded in job status. To enable
this feature, the user only needs to provide savepointsDir and set
fromLastSavepoint to true.

### CODE REVIEW

File: controllers/flinkcluster_reconciler.go [hunk 1]
```diff
@@ -338,30 +340,74 @@ func (reconciler *ClusterReconciler) reconcileJob() (ctrl.Result, error) {
...
+		}
+	}
 
 	// Create
 	if desiredJob != nil && observedJob == nil {
 		// If the observed Flink job status list is not nil (e.g., emtpy list), it
 		// means Flink REST API server is up and running. It is the source of
 		// truth of whether we can submit a job.
-		if reconciler.observed.flinkJobList != nil {
+		var jobs = reconciler.observed.flinkJobList
```
Comment (hongyegong) [inline controllers/flinkcluster_reconciler.go]: Probably k8s naming convention, I personally feel super confusing about jobs and status interchanging naming style. Might not be something that needs to be change in this PR, but would be great if we can mitigate this confusion in the future.

File: controllers/flinkclient/http_client.go [hunk 2]
```diff
@@ -52,10 +53,20 @@ func (c *HTTPClient) doHTTP(
...
+	if err != nil {
+		return err
+	}
 	resp, err := httpClient.Do(req)
 	if err != nil {
 		return err
 	}
+	c.Log.Info(
+		"HTTPClient", "status", resp.Status, "body", outStructPtr, "error", err)
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
```
[the remark this answers is not in the document]
Comment (functicons) [reply controllers/flinkclient/http_client.go]: Discussed offline, this is part of HTTP spec.

File: controllers/flinkcluster_observer.go [hunk 3]
```diff
@@ -231,16 +250,44 @@ func (observer *ClusterStateObserver) observeFlinkJobs(
...
-		log.Info("Observed Flink job status list", "jobs", jobs)
-		if jobCount > 1 {
+	if jobList == nil {
+		return
+	}
+
+	log.Info("Observed Flink job status list", "jobs", jobList.Jobs)
+	var flinkJobID *string
+	if len(jobList.Jobs) == 1 {
+		flinkJobID = &jobList.Jobs[0].ID
```
[the remark this answers is not in the document]
Comment (functicons) [reply controllers/flinkcluster_observer.go]: We want to find the current job here, if there is only one job, it is always the current, regardless of its status.

File: controllers/flinkcluster_observer.go [hunk 4]
```diff
@@ -231,16 +250,44 @@ func (observer *ClusterStateObserver) observeFlinkJobs(
...
+				"",
+				"jobs",
+				runningJobs)
+		}
+	}
+	if flinkJobID != nil {
+		var previousJobStatus = observed.cluster.Status.Components.PreviousJob
+		var specGen = observed.cluster.ObjectMeta.GetGeneration()
+		var statusGen = observed.cluster.Status.Generation
+		var isPreviousJob = (previousJobStatus != nil &&
```
[the remark this answers is not in the document]
Comment (functicons) [reply controllers/flinkcluster_observer.go]: The boolean is about "previous" not "observed".

File: controllers/flinkcluster_reconciler.go [hunk 5]
```diff
@@ -338,30 +340,74 @@ func (reconciler *ClusterReconciler) reconcileJob() (ctrl.Result, error) {
 	var desiredJob = reconciler.desired.Job
 	var observed = reconciler.observed
 	var observedJob = observed.job
+	var observedPrevJob = observed.previousJob
+	var observedPrevJobStatus = observed.cluster.Status.Components.PreviousJob
+	var runningJobs = []string{}
+
+	if reconciler.observed.flinkJobList != nil {
```
[the remark this answers is not in the document]
Comment (functicons) [reply controllers/flinkcluster_reconciler.go]: observed means observedState. we can replace reconciler.observed with observed.

**Comment by functicons** [comment]
Decided not to merge this PR.

**Pull request closed without merging**